composer/datasets/hparams.py

```diff
@@ -43,15 +49,116 @@ class DataloaderSpec(NamedTuple):
     split_fn: Callable[[Batch, int], List[Batch]] = _split_fn
 
 
-@dataclass
-class DatasetHparams(hp.Hparams, ABC):
-    """Abstract base class for hyperparameters to initialize a dataset."""
+@dataclasses.dataclass
+class DatadirHparamsMixin(hp.Hparams, abc.ABC):
+    """:attr:`datadir` field mixin for :class:`DatasetHparams`.
+
+    Parameters:
+        datadir (str): The path to the data directory.
+    """
+    datadir: Optional[str] = hp.optional("The path to the data directory", default=None)
+
+
+@dataclasses.dataclass
+class SyntheticBatchesHparamsMixin(hp.Hparams, abc.ABC):
+    """:attr:`synthetic` field mixin for :class:`DatasetHparams`.
+
+    Parameters:
+        synthetic (hp.Hparams, optional):
+            Parameters to use for synthetic data generation. If None (the default),
+            then real data will be used.
+    """
+
+    synthetic: Optional[hp.Hparams] = hp.optional(textwrap.dedent("""Parameters to use for synthetic data generation.
+        If None (the default), then real data will be used."""),
+                                                  default=None)
+
+    @classmethod
+    def get_synthetic_hparams_cls(cls) -> Type[hp.Hparams]:
+        """Returns the type of the dataclass for field :attr:`synthetic`.
+
+        Returns:
+            Type[hp.Hparams]: The type of the field :attr:`synthetic`.
+        """
+        field_types = get_type_hints(cls)
+        for field in dataclasses.fields(cls):
+            if field.name == "synthetic":
+                hparams_type = HparamsType(field_types[field.name])
+                return hparams_type.type
+        raise RuntimeError(f"Invariant error -- Dataset {cls.__name__} does not support synthetic data")
+
+
+@dataclasses.dataclass
+class NumTotalBatchesHparamsMixin(hp.Hparams, abc.ABC):
+    """:attr:`num_total_batches` field mixin for :class:`DatasetHparams`.
+
+    Parameters:
+        num_total_batches (int, optional):
+            If specified, then the dataloader from :meth:`initialize_object`
+            should yield this many batches per iteration. Specifically, ``len(dataloader) == num_total_batches``.
+            Each epoch should yield the same subset of samples.
+            
+            If this value is greater than the total number of samples in the dataset, then a :class:`ValueError` 
+            may be raised.
+
+            If None (the default), then the entire dataset will be iterated over.
+    """
+    num_total_batches: Optional[int] = hp.optional(
```

Comment:
Definitely am open to a better name!

The goal of this parameter is to support tests, smoketesting, and the profiler, where we would like to just do a few batches (these features view the world in terms of batches, not samples). 

`subset_examples` would work for images, though then these features would need to consider the `batch_size`, `eval_batch_size`, and `ddp.get_world_size()` when setting the `subset_examples`. When it's `num_total_batches`, it's really simple to hardcode it. I think this would also work for NLP, so long as the `batch_size` is in the same unit (samples or tokens) as the value of this field.

`subset_percent` would not work, as that would introduce floating point errors and assumes that we know the lenght of the dataset.

Reply:
Another question if we went with `subset_examples`: would this be the per-device # of examples, or the global number of examples across all ranks?

Reply:
Keep as is but rename to `subset_num_batches`